### PR TITLE
Prefer GroupVersionResource, fallback to GVK

### PR DESCRIPTION
**Pull request by juanvallejo**
**Release note**:
```release-note
NONE
```

Addresses https://github.com/kubernetes/kubernetes/pull/59353#discussion_r173048411

cc @smarterclayton @deads2k @soltysh

### CODE REVIEW

File: pkg/kubectl/resource/builder.go [hunk 1]
```diff
@@ -630,8 +614,23 @@ func (b *Builder) mappingFor(resourceOrKindArg string) (*meta.RESTMapping, error
...
+	}
+
+	if !fullySpecifiedGVK.Empty() {
+		if mapping, err := b.mapper.RESTMapping(fullySpecifiedGVK.GroupKind(), fullySpecifiedGVK.Version); err == nil {
+			return mapping, nil
+		}
+	}
 
-	return b.mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	return b.mapper.RESTMapping(groupKind, "")
```
Comment (deads2k) [inline pkg/kubectl/resource/builder.go]: if this errors out, you want to return `return b.mapper.RESTMapping(gvk.GroupKind(), gvk.Version)`, right?

Comment (juanvallejo) [reply pkg/kubectl/resource/builder.go]: Thanks, updated this

File: pkg/kubectl/resource/builder.go [hunk 2]
```diff
@@ -602,6 +602,22 @@ func (b *Builder) SingleResourceType() *Builder {
 // prefers a fully specified GroupVersionKind match. If we don't have one, match on a fully specified
 // GroupVersionResource, or fallback to a match on GroupResource.
```
Comment (soltysh) [inline pkg/kubectl/resource/builder.go]: Given the current changes the godoc needs update as well.

**Review by soltysh — APPROVED** [review]
One small nit, otherwise lgtm.

File: pkg/kubectl/resource/builder.go (hunk 2 above)
Comment (juanvallejo) [reply pkg/kubectl/resource/builder.go]: Thanks for catching that. Updated godoc to reflect new changes

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by soltysh** [comment]
/assign @smarterclayton 
for approval
/retest

File: hack/make-rules/test-cmd-util.sh [hunk 3]
```diff
@@ -3707,7 +3707,7 @@ run_RESTMapper_evaluation_tests() {
   # Pre-condition: None
   # Command
   kubectl get "${kube_flags[@]}" unknownresourcetype 2>${RESTMAPPER_ERROR_FILE} || true
-  if grep -q "the server doesn't have a resource type" "${RESTMAPPER_ERROR_FILE}"; then
+  if grep -q "no matches for kind \"unknownresourcetype\"" "${RESTMAPPER_ERROR_FILE}"; then
```
Comment (smarterclayton) [inline hack/make-rules/test-cmd-util.sh]: This message is technically incorrect - we couldn't find a resource or a kind called unknown resource type, and we are *very* careful in the CLI to describe this argument as a resource argument.  I think you need to fix the message in the case where you fall all the way through to be consistent with the previous release.

Comment (juanvallejo) [reply hack/make-rules/test-cmd-util.sh]: Thanks, added an error message when this happens to maintain consistency with previous behavior

**Review by soltysh — APPROVED** [review]
/lgtm

**Pull request merged**